### PR TITLE
Use ngx_posted_delayed_events implement 0-delay timers

### PR DESCRIPTION
**Pull request by syzh**
implement  #1766

### CODE REVIEW

**Review by spacewander — APPROVED** [review]

**Review by spacewander — CHANGES_REQUESTED** [review]
https://travis-ci.org/github/openresty/lua-nginx-module/jobs/722474171#L996
Need to update the test case.

**Comment by syzh** [comment]
test host vm 1c2g centos7
nginx 1 worker

### performance test data
**before** 
`ngx.sleep(0)`  avg: 6758.85
ab -c 10 -n 2000 http://localhost/t0   
1:
Server Software:        openresty/1.17.8.2
Server Hostname:        localhost
Server Port:            80

Document Path:          /t0
Document Length:        50 bytes

Concurrency Level:      10
Time taken for tests:   0.312 seconds
Complete requests:      2000
Failed requests:        225
   (Connect: 0, Receive: 0, Length: 225, Exceptions: 0)
Write errors:           0
Total transferred:      445530 bytes
HTML transferred:       99530 bytes
Requests per second:    6412.72 [#/sec] (mean)
Time per request:       1.559 [ms] (mean)
Time per request:       0.156 [ms] (mean, across all concurrent requests)
Transfer rate:          1395.05 [Kbytes/sec] received

2:
Server Software:        openresty/1.17.8.2
Server Hostname:        localhost
Server Port:            80

Document Path:          /t0
Document Length:        50 bytes

Concurrency Level:      10
Time taken for tests:   0.284 seconds
Complete requests:      2000
Failed requests:        206
   (Connect: 0, Receive: 0, Length: 206, Exceptions: 0)
Write errors:           0
Total transferred:      445574 bytes
HTML transferred:       99574 bytes
Requests per second:    7034.67 [#/sec] (mean)
Time per request:       1.422 [ms] (mean)
Time per request:       0.142 [ms] (mean, across all concurrent requests)
Transfer rate:          1530.50 [Kbytes/sec] received

3:
Server Software:        openresty/1.17.8.2
Server Hostname:        localhost
Server Port:            80

Document Path:          /t0
Document Length:        50 bytes

Concurrency Level:      10
Time taken for tests:   0.293 seconds
Complete requests:      2000
Failed requests:        205
   (Connect: 0, Receive: 0, Length: 205, Exceptions: 0)
Write errors:           0
Total transferred:      445576 bytes
HTML transferred:       99576 bytes
Requests per second:    6829.16 [#/sec] (mean)
Time per request:       1.464 [ms] (mean)
Time per request:       0.146 [ms] (mean, across all concurrent requests)
Transfer rate:          1485.79 [Kbytes/sec] received


`ngx.timer.at(0)`  avg: 5689.95
ab -c 10 -n 2000 http://localhost/timer   
1:
Server Software:        openresty/1.17.8.2
Server Hostname:        localhost
Server Port:            80

Document Path:          /timer
Document Length:        22 bytes

Concurrency Level:      10
Time taken for tests:   0.374 seconds
Complete requests:      2000
Failed requests:        233
   (Connect: 0, Receive: 0, Length: 233, Exceptions: 0)
Write errors:           0
Total transferred:      389724 bytes
HTML transferred:       43724 bytes
Requests per second:    5346.75 [#/sec] (mean)
Time per request:       1.870 [ms] (mean)
Time per request:       0.187 [ms] (mean, across all concurrent requests)
Transfer rate:          1017.46 [Kbytes/sec] received

2:
Server Software:        openresty/1.17.8.2
Server Hostname:        localhost
Server Port:            80

Document Path:          /timer
Document Length:        22 bytes

Concurrency Level:      10
Time taken for tests:   0.366 seconds
Complete requests:      2000
Failed requests:        223
   (Connect: 0, Receive: 0, Length: 223, Exceptions: 0)
Write errors:           0
Total transferred:      389752 bytes
HTML transferred:       43752 bytes
Requests per second:    5457.67 [#/sec] (mean)
Time per request:       1.832 [ms] (mean)
Time per request:       0.183 [ms] (mean, across all concurrent requests)
Transfer rate:          1038.64 [Kbytes/sec] received

3:
Server Software:        openresty/1.17.8.2
Server Hostname:        localhost
Server Port:            80

Document Path:          /timer
Document Length:        22 bytes

Concurrency Level:      10
Time taken for tests:   0.319 seconds
Complete requests:      2000
Failed requests:        201
   (Connect: 0, Receive: 0, Length: 201, Exceptions: 0)
Write errors:           0
Total transferred:      389774 bytes
HTML transferred:       43774 bytes
Requests per second:    6265.43 [#/sec] (mean)
Time per request:       1.596 [ms] (mean)
Time per request:       0.160 [ms] (mean, across all concurrent requests)
Transfer rate:          1192.43 [Kbytes/sec] received

**after:** 
`ngx.timer.at(0)` avg: 6997.50
ab -c 10 -n 2000 http://localhost/timer   
Server Software:        openresty/1.17.8.2
Server Hostname:        localhost
Server Port:            80

Document Path:          /timer
Document Length:        22 bytes

Concurrency Level:      10
Time taken for tests:   0.286 seconds
Complete requests:      2000
Failed requests:        219
   (Connect: 0, Receive: 0, Length: 219, Exceptions: 0)
Write errors:           0
Total transferred:      389743 bytes
HTML transferred:       43743 bytes
Requests per second:    6989.17 [#/sec] (mean)
Time per request:       1.431 [ms] (mean)
Time per request:       0.143 [ms] (mean, across all concurrent requests)
Transfer rate:          1330.07 [Kbytes/sec] received


Server Software:        openresty/1.17.8.2
Server Hostname:        localhost
Server Port:            80

Document Path:          /timer
Document Length:        22 bytes

Concurrency Level:      10
Time taken for tests:   0.289 seconds
Complete requests:      2000
Failed requests:        214
   (Connect: 0, Receive: 0, Length: 214, Exceptions: 0)
Write errors:           0
Total transferred:      389773 bytes
HTML transferred:       43773 bytes
Requests per second:    6927.63 [#/sec] (mean)
Time per request:       1.443 [ms] (mean)
Time per request:       0.144 [ms] (mean, across all concurrent requests)
Transfer rate:          1318.46 [Kbytes/sec] received

Server Software:        openresty/1.17.8.2
Server Hostname:        localhost
Server Port:            80

Document Path:          /timer
Document Length:        22 bytes

Concurrency Level:      10
Time taken for tests:   0.283 seconds
Complete requests:      2000
Failed requests:        186
   (Connect: 0, Receive: 0, Length: 186, Exceptions: 0)
Write errors:           0
Total transferred:      389780 bytes
HTML transferred:       43780 bytes
Requests per second:    7075.71 [#/sec] (mean)
Time per request:       1.413 [ms] (mean)
Time per request:       0.141 [ms] (mean, across all concurrent requests)
Transfer rate:          1346.67 [Kbytes/sec] received


`ngx.sleep(0)` avg: 6986.73
ab -c 10 -n 2000 http://localhost/t0

Server Software:        openresty/1.17.8.2
Server Hostname:        localhost
Server Port:            80

Document Path:          /t0
Document Length:        48 bytes

Concurrency Level:      10
Time taken for tests:   0.266 seconds
Complete requests:      2000
Failed requests:        1838
   (Connect: 0, Receive: 0, Length: 1838, Exceptions: 0)
Write errors:           0
Total transferred:      445556 bytes
HTML transferred:       99556 bytes
Requests per second:    7525.19 [#/sec] (mean)
Time per request:       1.329 [ms] (mean)
Time per request:       0.133 [ms] (mean, across all concurrent requests)
Transfer rate:          1637.16 [Kbytes/sec] received


Server Software:        openresty/1.17.8.2
Server Hostname:        localhost
Server Port:            80

Document Path:          /t0
Document Length:        50 bytes

Concurrency Level:      10
Time taken for tests:   0.293 seconds
Complete requests:      2000
Failed requests:        192
   (Connect: 0, Receive: 0, Length: 192, Exceptions: 0)
Write errors:           0
Total transferred:      445578 bytes
HTML transferred:       99578 bytes
Requests per second:    6824.94 [#/sec] (mean)
Time per request:       1.465 [ms] (mean)
Time per request:       0.147 [ms] (mean, across all concurrent requests)
Transfer rate:          1484.88 [Kbytes/sec] received

Server Software:        openresty/1.17.8.2
Server Hostname:        localhost
Server Port:            80

Document Path:          /t0
Document Length:        50 bytes

Concurrency Level:      10
Time taken for tests:   0.303 seconds
Complete requests:      2000
Failed requests:        202
   (Connect: 0, Receive: 0, Length: 202, Exceptions: 0)
Write errors:           0
Total transferred:      445536 bytes
HTML transferred:       99536 bytes
Requests per second:    6610.08 [#/sec] (mean)
Time per request:       1.513 [ms] (mean)
Time per request:       0.151 [ms] (mean, across all concurrent requests)
Transfer rate:          1438.00 [Kbytes/sec] received


code:
```
location /timer {
            content_by_lua_block{
                ngx.say('timer: ',  ngx.now())
            }
            log_by_lua_block{
                local function f1(p, uri)
                end
                local ok, err = ngx.timer.at(0, f1, ngx.var.uri)
                if not ok then
                    ngx.log(ngx.ERR, "failed to create timer: ", err)
                    return
                end

            }

        }

location /t0 {
            content_by_lua_block {
                ngx.sleep(0)
                ngx.say('hello-01: ',  ngx.now())
                ngx.sleep(0)
                ngx.say('world-02: ',  ngx.now())
            }
        }
```

File: src/ngx_http_lua_sleep.c [hunk 1]
```diff
@@ -71,7 +71,6 @@ ngx_http_lua_ngx_sleep(lua_State *L)
 #ifdef HAVE_POSTED_DELAYED_EVENTS_PATCH
         dd("posting 0 sec sleep event to head of delayed queue");
 
-        coctx->sleep.delayed = 1;
```
Comment (doujiang24) [inline src/ngx_http_lua_sleep.c:74]: I remember this is useful, @dndx can you help to confirm it?

File: src/ngx_http_lua_timer.c [hunk 2]
```diff
@@ -351,10 +351,20 @@ ngx_http_lua_ngx_timer_helper(lua_State *L, int every)
 
     lmcf->pending_timers++;
 
-    ngx_add_timer(ev, delay);
+    if (delay == 0) {
```
Comment (doujiang24) [inline src/ngx_http_lua_timer.c:354]: I personally like this form.
```
#ifdef HAVE_POSTED_DELAYED_EVENTS_PATCH
    if (delay == 0) {
        dd("posting 0 sec sleep event to head of delayed queue");
        ngx_post_event(ev, &ngx_posted_delayed_events);
        lua_pushinteger(L, 1);
        return 1;
    }
#endif
```

File: src/ngx_http_lua_sleep.c (hunk 1 above)
Comment (dndx) [reply src/ngx_http_lua_sleep.c:74]: Yes, this flag is used by the `HAVE_POSTED_DELAYED_EVENTS_PATCH` to signal that the event should be executed the next iteration. Without it the sleep will be executed immediately in the current iteration.

**Comment by dndx** [comment]
Actually I think @agentzh is correct, we could switch all the usage of `HAVE_POSTED_DELAYED_EVENTS_PATCH` to use

`ngx_post_event(ev, & ngx_posted_next_events);` instead, it will achieve the exact same effect and allow us to eliminate the patch completely.

File: src/ngx_http_lua_sleep.c (hunk 1 above)
Comment (syzh) [reply src/ngx_http_lua_sleep.c:74]: I know, but the next iteration will have one more epoll_wait(0) system call

Comment (doujiang24) [reply src/ngx_http_lua_sleep.c:74]: we need to make sure the behavior is expected.

File: t/077-sleep.t [hunk 3]
```diff
@@ -438,18 +438,12 @@ ok
 qr/f begin
 f begin
 f begin
-worker cycle
```
Comment (doujiang24) [inline t/077-sleep.t:441]: we need to reset these test cases.

**Review by doujiang24 — COMMENTED** [review]
otherwise, LGTM.

**Review by dndx — APPROVED** [review]
Aside from the test failure, this change LGTM.

**Comment by doujiang24** [comment]
@syzh Merged, thanks 👍

**Pull request closed without merging**